File: problems/leetcode/1100/1106.cpp
// 1106. Parsing A Boolean Expression
// https://leetcode.com/problems/parsing-a-boolean-expression/description/

using namespace std;

#include <string>

class Solution {
public:
  pair<int, bool> parseBoolExprRecur(string &expression, int index) {
    switch (expression[index]) {
    case 't':
      return {index + 1, true};
    case 'f':
      return {index + 1, false};
    case '!': {
      pair<int, bool> innerResult = parseBoolExprRecur(expression, index + 2);
      return {innerResult.first + 1, !innerResult.second};
    }
    case '&': {
      int curr = index + 2;
      bool result = true;

      while (expression[curr] != ')') {
        pair<int, bool> innerResult = parseBoolExprRecur(expression, curr);
        result = result && innerResult.second;
        curr = innerResult.first;
        if (curr < expression.size() && expression[curr] == ',') {
          curr++;
        }
      }

      return {curr + 1, result};
    }
    case '|': {
      int curr = index + 2;
      bool result = false;

      while (expression[curr] != ')') {
        pair<int, bool> innerResult = parseBoolExprRecur(expression, curr);
        result = result || innerResult.second;
        curr = innerResult.first;
        if (curr < expression.size() && expression[curr] == ',') {
          curr++;
        }
      }

      return {curr + 1, result};
    }
    default:
      break;
    }

    return {0, false};
  }

  bool parseBoolExpr(string expression) {
    return parseBoolExprRecur(expression, 0).second;
  }
};

File: problems/leetcode/1100/1106/1106.py
# 1106. Parsing A Boolean Expression
# https://leetcode.com/problems/parsing-a-boolean-expression/description

class Solution:
    def parseBoolExpr(self, expression: str) -> bool:
        n = len(expression)
        return self.parseBoolExprRecurr(expression, 0)[0]

    def parseBoolExprRecurr(self, expression: str, left: str) -> tuple[bool, int]:
        if expression[left] == "t":
            return True, left + 1
        elif expression[left] == "f":
            return False, left + 1
        elif expression[left] == "!":
            result, right = self.parseBoolExprRecurr(expression, left + 2)
            return not result, right + 1
        elif expression[left] in ["&", "|"]:
            result = True if expression[left] == "&" else False
            index = left + 1

            while expression[index] != ")":
                index += 1
                innerResult, index = self.parseBoolExprRecurr(expression, index)
                result = result and innerResult if expression[left] == "&" else result or innerResult

            return result, index + 1
